### PR TITLE
Add Kotest and Programmatic support

### PR DESCRIPTION
**Pull request by Drakojin**

### CODE REVIEW

**Comment by Drakojin** [comment]
Update for documentation is still missing right now. But i would appreciate an early code review

File: logrecorder-jul/src/main/kotlin/io/github/logrecorder/jul/kotest/JulRecorderKotestExtension.kt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+    vararg classes: KClass<*>,
+    names: Array<out String> = emptyArray()
+) = JulRecorderKotestExtension(classes, names)
+
+/**
+ * @see recordLogs
+ * @since 2.4
+ */
+@PublishedApi
+internal fun recordLogs(
```
Comment (slu-it) [inline logrecorder-jul/src/main/kotlin/io/github/logrecorder/jul/kotest/JulRecorderKotestExtension.kt:35]: if they are all `internal`, nobody can use them - how do I use the extension?

File: logrecorder-log4j/src/main/kotlin/io/github/logrecorder/log4j/kotest/Log4jRecorderKotestExtension.kt [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+ * Recording a logger will set its log level to [org.apache.logging.log4j.Level.ALL] for the duration of the test.
+ * After the test was executed, the log level will be restored to whatever it was before.
+ *
+ * @param classes [KClass] whose name should be used to identify loggers to record.
+ * @param names String names of the [Logger] instances to record
+ * @see Log4jRecorderKotestExtension
+ * @since 2.4
+ */
+@PublishedApi
+internal fun recordLogs(
```
Comment (slu-it) [inline logrecorder-log4j/src/main/kotlin/io/github/logrecorder/log4j/kotest/Log4jRecorderKotestExtension.kt:25]: if they are all `internal`, nobody can use them - how do I use the extension?

File: logrecorder-logback/src/main/kotlin/io/github/logrecorder/logback/kotest/LogbackRecorderKotestExtension.kt [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ * Recording a logger will set its log level to [Level.ALL] for the duration of the test.
+ * After the test was executed, the log level will be restored to whatever it was before.
+ *
+ * @param classes [KClass] whose name should be used to identify loggers to record.
+ * @param names String names of the [Logger] instances to record
+ * @see LogbackRecorderKotestExtension
+ * @since 2.4
+ */
+@PublishedApi
+internal fun recordLogs(
```
Comment (slu-it) [inline logrecorder-logback/src/main/kotlin/io/github/logrecorder/logback/kotest/LogbackRecorderKotestExtension.kt:26]: if they are all `internal`, nobody can use them - how do I use the extension?

**Review by slu-it — CHANGES_REQUESTED** [review]

File: logrecorder-common/src/main/kotlin/io/github/logrecorder/common/LogRecorderExecutionBase.kt [hunk 4]
```diff
@@ -0,0 +1,19 @@
+package io.github.logrecorder.common
+
+import io.github.logrecorder.api.MutableLogRecord
+import kotlin.reflect.KClass
+
+abstract class LogRecorderExecutionBase<L : Any, LR : MutableLogRecord<*>> {
+    abstract val loggerFromKClass: (KClass<*>) -> L
+    abstract val loggerFromName: (String) -> L
```
[the remark this answers is not in the document]
Comment (Drakojin) [reply logrecorder-common/src/main/kotlin/io/github/logrecorder/common/LogRecorderExecutionBase.kt:8]: I don't get it. Do you have an example?

File: logrecorder-jul/src/main/kotlin/io/github/logrecorder/jul/kotest/JulRecorderKotestExtension.kt (hunk 1 above)
Comment (Drakojin) [reply logrecorder-jul/src/main/kotlin/io/github/logrecorder/jul/kotest/JulRecorderKotestExtension.kt:35]: The @PublishedApi makes it usable

File: logrecorder-log4j/src/main/kotlin/io/github/logrecorder/log4j/kotest/Log4jRecorderKotestExtension.kt (hunk 2 above)
Comment (Drakojin) [reply logrecorder-log4j/src/main/kotlin/io/github/logrecorder/log4j/kotest/Log4jRecorderKotestExtension.kt:25]: The @PublishedApi makes it usable

File: logrecorder-logback/src/main/kotlin/io/github/logrecorder/logback/kotest/LogbackRecorderKotestExtension.kt (hunk 3 above)
Comment (Drakojin) [reply logrecorder-logback/src/main/kotlin/io/github/logrecorder/logback/kotest/LogbackRecorderKotestExtension.kt:26]: The @PublishedApi makes it usable

**Review by slu-it — APPROVED** [review]

**Pull request merged**